tunnel/intra/doh/client_auth_test.go

```diff
@@ -0,0 +1,378 @@
+FL9DGpUBsYe03F5eZ/lJikopL3voqKDCJQKKgJk0jb0jXjwAgQ86TX+G+hezL5jp
+xOOfMmTYgMwnUuFYN1gHAd+TnYB9G1qSQr9TOw3K9Rf4q2x09GhLP75qdr+qzmIR
+YGle5ZSP0LqKNkpGNUECQQD+6CxOO8+knnzIFvqkUyNDVFR5ALRNpb53TGVITNf3
+ysT32oJ75ButA0l4q/jsL+MeLLvrHkJOHN+ydLaZOUkbAkEA3m5cICisW9lsT+Rj
+glXykkbj3Ougldy7rhPivAaS7clk8cl8cDcIvHna1mDlhSanUu/s4TFEXBLnSzee
+XLNIcQJBAJ0n3TD6lSEkCUB/UlX/X81B77aOZZs9pXj9o6/4mGoQHHHGyQ3C7AE1
+9pUsSZKsT3UqFU124WAxUwU+CdnbxKMCQB/QrUC0UKL6oHF0+37DCGU/2ovY8Ck/
+X2Dw2zeFwTJd4iBrb28lkAxVaaXMSkgXVUuZoco8H8kDsy2hEPe1dSECQQCPw5Yg
+2gdmdpUk+QetqqhSuuIDwILHU9m3CoX3rY+njaR5LOWDz3utC9Ogo+4wdIMamP/o
+2SAWPAZPqDUbtqGH
+-----END PRIVATE KEY-----`
+
+// fakeClientAuth implements the ClientAuth interface for testing.
+type fakeClientAuth struct {
+	ClientAuth
+
+	certificate  *x509.Certificate
+	intermediate *x509.Certificate
+	key          crypto.PrivateKey
+}
+
+func (ca *fakeClientAuth) GetClientCertificate() []byte {
+	if ca.certificate == nil {
+		// Interface uses nil for errors to support binding.
+		return nil
+	}
+	return ca.certificate.Raw
+}
+
+func (ca *fakeClientAuth) GetIntermediateCertificate() []byte {
+	if ca.intermediate == nil {
+		return nil
+	}
+	return ca.intermediate.Raw
+}
+
+func (ca *fakeClientAuth) Sign(digest []byte) []byte {
+	if ca.key == nil {
+		return nil
+	}
+	if k, isECDSA := ca.key.(*ecdsa.PrivateKey); isECDSA {
+		// We have to ASN.1 encode the signature ourselves since
+		// Go < 1.15 does not include ecdsa.SignASN1
+		// (https://github.com/golang/go/issues/20544).
+		r, s, err := ecdsa.Sign(rand.Reader, k, digest)
+		if err != nil {
+			return nil
+		}
+
+		var b cryptobyte.Builder
+		b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+			b.AddASN1BigInt(r)
+			b.AddASN1BigInt(s)
+		})
+		signature, err := b.Bytes()
+		if err != nil {
+			return nil
+		}
+		return signature
+	}
+	// Unsupported key type
+	return nil
+}
+
+func newFakeClientAuth(certificate, intermediate, key []byte) (*fakeClientAuth, error) {
+	ca := &fakeClientAuth{}
+	if certificate != nil {
+		certX509, err := x509.ParseCertificate(certificate)
+		if err != nil {
+			return nil, fmt.Errorf("certificate: %v", err)
+		}
+		ca.certificate = certX509
+	}
+	if intermediate != nil {
+		intX509, err := x509.ParseCertificate(intermediate)
+		if err != nil {
+			return nil, fmt.Errorf("intermediate: %v", err)
+		}
+		ca.intermediate = intX509
+	}
+	if key != nil {
+		key, err := x509.ParsePKCS8PrivateKey(key)
+		if err != nil {
+			return nil, fmt.Errorf("private key: %v", err)
+		}
+		ca.key = key
+	}
+	return ca, nil
+}
+
+func newCertificateRequestInfo() *tls.CertificateRequestInfo {
+	return &tls.CertificateRequestInfo{
+		Version: tls.VersionTLS13,
+	}
+}
+
+func newToBeSigned(message []byte) ([]byte, crypto.SignerOpts) {
+	digest := sha256.Sum256(message)
+	opts := crypto.SignerOpts(crypto.SHA256)
+	return digest[:], opts
+}
+
+// VerifyASN1 handles ECDSA ASN.1 signatures for Go < 1.15 support.
```

Comment:
Can we remove this function now that we require go 1.15?

Reply:
Excellent.